cache/contenthash/checksum_test.go

```diff
@@ -0,0 +1,473 @@
+
+	ch = []string{
+		"DEL d0 dir",
+	}
+
+	err = emit(cc.HandleChange, changeStream(ch))
+	assert.NoError(t, err)
+
+	_, err = cc.Checksum(context.TODO(), ref, "d0")
+	assert.Error(t, err)
+	assert.Equal(t, errNotFound, errors.Cause(err))
+
+	_, err = cc.Checksum(context.TODO(), ref, "d0/abc")
+	assert.Error(t, err)
+	assert.Equal(t, errNotFound, errors.Cause(err))
+
+	err = ref.Release(context.TODO())
+	require.NoError(t, err)
+}
+
+func TestPersistence(t *testing.T) {
+	tmpdir, err := ioutil.TempDir("", "buildkit-state")
+	require.NoError(t, err)
+	defer os.RemoveAll(tmpdir)
+
+	cm := setupCacheManager(t, tmpdir)
+	defer cm.Close()
+
+	ch := []string{
+		"ADD foo file data0",
+		"ADD bar file data1",
+		"ADD d0 dir",
+		"ADD d0/abc file data0",
+		"ADD d0/def symlink abc",
+		"ADD d0/ghi symlink nosuchfile",
+	}
+
+	ref := createRef(t, cm, ch)
+	id := ref.ID()
+
+	dgst, err := Checksum(context.TODO(), ref, "foo")
+	assert.NoError(t, err)
+	assert.Equal(t, dgstFileData0, dgst)
+
+	err = ref.Release(context.TODO())
+	require.NoError(t, err)
+
+	ref, err = cm.Get(context.TODO(), id)
+	assert.NoError(t, err)
+
+	dgst, err = Checksum(context.TODO(), ref, "foo")
+	assert.NoError(t, err)
+	assert.Equal(t, dgstFileData0, dgst)
+
+	err = ref.Release(context.TODO())
+	require.NoError(t, err)
+
+	time.Sleep(100 * time.Millisecond) // saving happens on the background
+
+	cm.Close()
+	getDefaultManager().lru.Purge()
+	cm = setupCacheManager(t, tmpdir)
+	defer cm.Close()
+
+	ref, err = cm.Get(context.TODO(), id)
+	assert.NoError(t, err)
+
+	dgst, err = Checksum(context.TODO(), ref, "foo")
+	assert.NoError(t, err)
+	assert.Equal(t, dgstFileData0, dgst)
+}
+
+func createRef(t *testing.T, cm cache.Manager, files []string) cache.ImmutableRef {
+	mref, err := cm.New(context.TODO(), nil, cache.CachePolicyRetain)
+	require.NoError(t, err)
+
+	mounts, err := mref.Mount(context.TODO(), false)
+	require.NoError(t, err)
+
+	lm := snapshot.LocalMounter(mounts)
+
+	mp, err := lm.Mount()
+	require.NoError(t, err)
+
+	err = writeChanges(mp, changeStream(files))
+	lm.Unmount()
+	require.NoError(t, err)
+
+	ref, err := mref.Commit(context.TODO())
+	require.NoError(t, err)
+
+	return ref
+}
+
+func setupCacheManager(t *testing.T, tmpdir string) cache.Manager {
+	snapshotter, err := naive.NewSnapshotter(filepath.Join(tmpdir, "snapshots"))
+	require.NoError(t, err)
+
+	md, err := metadata.NewStore(filepath.Join(tmpdir, "metadata.db"))
+	require.NoError(t, err)
+
+	cm, err := cache.NewManager(cache.ManagerOpt{
+		Snapshotter:   snapshotter,
+		MetadataStore: md,
+	})
+	require.NoError(t, err)
+
+	return cm
+}
+
+// these test helpers are from tonistiigi/fsutil
+
+type change struct {
+	kind fsutil.ChangeKind
+	path string
+	fi   os.FileInfo
+	data string
+}
+
+func changeStream(dt []string) (changes []*change) {
+	for _, s := range dt {
+		changes = append(changes, parseChange(s))
+	}
+	return
+}
+
+func parseChange(str string) *change {
```

Comment:
I think the test reads slightly better. Things like size calculation and mode fixes can't be done with inline structure. An alternative would be to define a helper function for file, dir and symlink. Same helper function is in fsutil.